### ISSUE TITLE
Consider replacing rebind_executor trait

### ISSUE THREAD

**Issue by jaredhoberock** [jaredhoberock/agency#218]
The way that the `rebind_executor` trait works is kind of confusing.

An execution policy's `.on()` function uses it to return a new execution policy with the given type rebound. For example, `detail::basic_execution_policy::on()` does this:

    template<class OtherExecutor>
    rebind_executor_t<derived_type, OtherExecutor> on(const OtherExecutor& executor) const
    {
      return rebind_executor_t<derived_type, OtherExecutor>(param(), executor);
    }

That means if we want to do something special, we need to specialize `rebind_executor_t` on the type of execution policy (`derived_type`) we want to modify, as well as the type of executor (`OtherExecutor`) we want to rebind. We also have to ensure that the constructor of the execution policy type named by `rebind_executor_t` knows what to do with `param()` and `executor`.

It might be simpler to provide a customization point `replace_executor()`. This would be a function that returns an execution policy and could be specialized by overloading. With `replace_executor()`, `detail::basic_execution_policy::on()` would look like this:

    template<class OtherExecutor>
    auto on(const OtherExecutor& exec) const
    {
      using agency::replace_executor;
      return replace_executor(*this, exec);
    }

In other words, `.on()` is just sugar for `replace_executor()`.
